### ISSUE TITLE
Merge UniValue refactoring from Bitcoin

### ISSUE THREAD

**Issue by str4d** [zcash/zcash#1985]
Requires cherry-picking bitcoin/bitcoin#6121 and any other UniValue-related improvements since 0.11.2.

This will be a bunch of work, but is a blocker for pulling various newer patches from upstream (see e.g. #1593).

**Comment by bitcartel** [comment]
Regarding ticket #1593, it was resolved and I don't believe any new tickets were filed about the RPC interface locking up.  If we want to restore connection keep-alive, debugging usage of boost::asio for the rpc server mght be sufficient.

Merging UniValue was discussed at an engineering meeting, but I don't believe a decision was reached.  Whether or not UniValue is merged, pulling in upstream patches will still require cleaning up, since segwit has touched many parts of the system.

Adding 'needs prioritization' label for @nathan-at-least

**Comment by str4d** [comment]
@bitcartel the primary reason/motivator for #1593 for me now is Tor's ephemeral Hidden Services, the upstream PR for which heavily uses libevent, and would require considerable engineering effort to reimplement ourselves.

Regarding this issue, I'm hacking on it this evening - I'm about half-way through cherry-picking the upstream PR, and then it just needs testing to spot and fix the changes we've made that use json::spirit. It's definitely going more smoothly than I anticipated.
